pkg/core/config/newtenant.go

```diff
@@ -0,0 +1,532 @@
+		return errors.New("DATABASE_URL is not set")
+	}
+	if c.AppName == "" {
+		return errors.New("APP_NAME is not set")
+	}
+	if c.UserConfig.APIKey == "" {
+		return errors.New("API_KEY is not set")
+	}
+	if c.UserConfig.MasterKey == "" {
+		return errors.New("MASTER_KEY is not set")
+	}
+	if c.UserConfig.APIKey == c.UserConfig.MasterKey {
+		return errors.New("MASTER_KEY cannot be the same as API_KEY")
+	}
+	return name.ValidateAppName(c.AppName)
+}
+
+func (c *TenantConfiguration) AfterUnmarshal() {
+	// Default token secret to master key
+	if c.UserConfig.TokenStore.Secret == "" {
+		c.UserConfig.TokenStore.Secret = c.UserConfig.MasterKey
+	}
+
+	// Propagate AppName
+	if c.UserConfig.ForgotPassword.AppName == "" {
+		c.UserConfig.ForgotPassword.AppName = c.AppName
+	}
+
+	// Propagate URLPrefix
+	if c.UserConfig.ForgotPassword.URLPrefix == "" {
+		c.UserConfig.ForgotPassword.URLPrefix = c.UserConfig.URLPrefix
+	}
+	if c.UserConfig.WelcomeEmail.URLPrefix == "" {
+		c.UserConfig.WelcomeEmail.URLPrefix = c.UserConfig.URLPrefix
+	}
+	if c.UserConfig.SSO.URLPrefix == "" {
+		c.UserConfig.SSO.URLPrefix = c.UserConfig.URLPrefix
+	}
+	if c.UserConfig.UserVerification.URLPrefix == "" {
+		c.UserConfig.UserVerification.URLPrefix = c.UserConfig.URLPrefix
+	}
+
+	// Remove trailing slash in URLs
+	c.UserConfig.URLPrefix = removeTrailingSlash(c.UserConfig.URLPrefix)
+	c.UserConfig.ForgotPassword.URLPrefix = removeTrailingSlash(c.UserConfig.ForgotPassword.URLPrefix)
+	c.UserConfig.WelcomeEmail.URLPrefix = removeTrailingSlash(c.UserConfig.WelcomeEmail.URLPrefix)
+	c.UserConfig.UserVerification.URLPrefix = removeTrailingSlash(c.UserConfig.UserVerification.URLPrefix)
+	c.UserConfig.SSO.URLPrefix = removeTrailingSlash(c.UserConfig.SSO.URLPrefix)
+}
+
+func GetTenantConfig(r *http.Request) TenantConfiguration {
+	s := r.Header.Get(coreHttp.HeaderTenantConfig)
+	config, err := NewTenantConfigurationFromStdBase64Msgpack(s)
+	if err != nil {
+		panic(err)
+	}
+	return *config
+}
+
+func SetTenantConfig(r *http.Request, config *TenantConfiguration) {
+	value, err := config.StdBase64Msgpack()
+	if err != nil {
+		panic(err)
+	}
+	r.Header.Set(coreHttp.HeaderTenantConfig, value)
+}
+
+func DelTenantConfig(r *http.Request) {
+	r.Header.Del(coreHttp.HeaderTenantConfig)
+}
+
+func removeTrailingSlash(url string) string {
+	if strings.HasSuffix(url, "/") {
+		return url[:len(url)-1]
+	}
+
+	return url
+}
+
+// UserConfiguration represents user-editable configuration
+type UserConfiguration struct {
+	APIKey           string                         `json:"api_key" yaml:"api_key" msg:"api_key"`
+	MasterKey        string                         `json:"master_key" yaml:"master_key" msg:"master_key"`
+	URLPrefix        string                         `json:"url_prefix" yaml:"url_prefix" msg:"url_prefix"`
+	CORS             CORSConfiguration              `json:"cors" yaml:"cors" msg:"cors"`
+	Auth             NewAuthConfiguration           `json:"auth" yaml:"auth" msg:"auth"`
```

Comment:
Do we remove the `New` prefix? The old configuration seems deleted.

Reply:
Yes. we should. Let me remove all `New` prefixes.